### PR TITLE
Dev.indices

### LINKED ISSUE #157: Error in indicies when using subscripts
converting `abcmn` though the mapping below produces incorrect indicies:
```
a,d
bc,e
g{1}h{2}i{3},G{2}H{1}I{3}J{1}
m{1}n{2},N{2}M{1}
```

To reproduce: `git checkout dev.compose-bug` (or `git checkout 9d3268f03e`) and run
```
cd g2p/tests
g2p convert --config  public/mappings/compose.yaml abcmn c2 c3 -d -e
```
current output:
```
[   'deNM',
    [['a', 'd'], ['b', 'e'], ['c', 'e'], ['c', 'M'], ['m', 'N']],
    [(0, 0), (1, 1), (2, 1), (2, 3), (3, 2)],
...
```
expected output:
```
[   'deNM',
    [['a', 'd'], ['b', 'e'], ['c', 'e'], ['m', 'M'], ['n', 'N']],
    [(0, 0), (1, 1), (2, 1), (3, 3), (4, 2)],
...
```
Note that converting just `mn` outputs correct indices, and so does converting `xnm`, the bug comes out when `nm` is preceded by `bc`.

### PR DESCRIPTION
**Pull request by roedoejet**
This is a PR stub for the fix for indices - it's not ready yet, so no need to review. I will change the status and add reviewers when ready. Closes #157 #172 and #173

### CODE REVIEW

**Comment by joanise** [comment]
There's a new warning that this PR causes in unit testing. It's in `python run.py dev` but more narrowly in `python test_cli.py`:
```
test_update (g2p.tests.test_cli.CliTest) ... WARNING - Failed to read language cache from /tmp/tmpzwfarfor/langs.pkl: unpickling stack underflow
WARNING - Failed to read language network from /tmp/tmpzwfarfor/network.pkl: could not find MARK
```
It occurs both on my machine and in CI. I don't know what this warning means, but it worries me a bit.

**Review by joanise — COMMENTED** [review]
This first review is just about commit fd2bb0fbc1 fixing `win` normalization, where I have a few questions. I'll review separately the indices work.

File: g2p/mappings/langs/generated/win-ipa_to_eng-ipa.json [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    {"in": "eː", "out": "eː"},
+    {"in": "iː", "out": "eː"},
+    {"in": "iː", "out": "eː"},
+    {"in": "ĩː", "out": "ẽː"},
+    {"in": "ĩː", "out": "ẽː"},
+    {"in": "dʒ", "out": "dʒ"},
+    {"in": "kʼ", "out": "k"},
+    {"in": "oː", "out": "oː"},
+    {"in": "oː", "out": "oː"},
+    {"in": "pʼ", "out": "p"},
```
[the remark this answers is not in the document]
Comment (joanise) [reply g2p/mappings/langs/generated/win-ipa_to_eng-ipa.json:17]: Hum, you changed win->win_ipa to map `t'` to `t`+U+02BC which Panphon says is the ejective marker, not a glottal stop, so maybe that is indeed the better choice.

**Comment by joanise** [comment]
> There's a new warning that this PR causes in unit testing. It's in `python run.py dev` but more narrowly in `python test_cli.py`:
> 
> ```
> test_update (g2p.tests.test_cli.CliTest) ... WARNING - Failed to read language cache from /tmp/tmpzwfarfor/langs.pkl: unpickling stack underflow
> WARNING - Failed to read language network from /tmp/tmpzwfarfor/network.pkl: could not find MARK
> ```
> 
> It occurs both on my machine and in CI. I don't know what this warning means, but it worries me a bit.

Never mind that problem. They're warnings caused deliberately by tests that corrupt langs.pkl and network.pkl deliberately and validates that our code recovers cleanly. They have nothing to do with this PR!

File: g2p/mappings/langs/generated/win-ipa_to_eng-ipa.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (joanise) [reply g2p/mappings/langs/generated/win-ipa_to_eng-ipa.json:17]: I've just added the test cases I suggested, please make sure they're correct.

File: g2p/transducer/__init__.py [hunk 2]
```diff
@@ -723,20 +823,20 @@ def check(
...
+                tg._tiers[i],
+                display_warnings=display_warnings,
+                original_input=tg.input_string,
+            ):
+                # Don't short circuit if warnings are required
+                if display_warnings:
+                    result = False
+                else:
+                    return False
+        return result
```
[the remark this answers is not in the document]
Comment (joanise) [reply g2p/transducer/__init__.py:839]: I'm sorry, but I strongly disagree with this change. The pylint warning `no-else-return` apparently comes from the Mozilla coding style guide (at least according to https://stackoverflow.com/a/55561880) but I think it should be disabled. I have disabled it in `g2p/.pylintrc` with `disable=no-else-return`, but maybe VSCode doesn't use that config file. Would you mind configuring your VSCode to disable it globally?

This blog post sums up my position on this very well: https://dmerej.info/blog/post/else-after-return-yea-or-nay/ (found at https://stackoverflow.com/a/57539298):
 - When you have a parallel structure where one block is "if a" and the other block if conceptually "if not a", and both blocks return, the `else` displays that visually to the programmer. While the else might be redundant to the interpreter, cognitively it is not redundant at all: it communicates the intent of the code to the next programmer reading the code.
 - When you have a guard block, like "if bad situation: return" then it makes sense not to use an else, because the rest of the code is not actually in a parallel structure, it's just unable to proceed when the guard fails. So an else here would distract the programmer reading the code rather than add clarity.

In this case, I believe you actually agreed with me, because you felt the need to add the comment "if not shallow, go deeper", which tells me that you thought following pylint's suggestion made the code less clear.

**Review by joanise — CHANGES_REQUESTED** [review]
Wow, this truly was index hell!!!

I'm very glad you wrote good unit test cases for this, because it's hard to follow the logic, but I'm confident the code is good (or at least it covers the cases we've discussed) because you're exercising them.

I've pushed a couple small changes, and there's one more style change I'm requesting, and several comments interspersed in the code, but great work. My sympathies for having had to do it!

File: g2p/mappings/langs/generated/win-ipa_to_eng-ipa.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (roedoejet) [reply g2p/mappings/langs/generated/win-ipa_to_eng-ipa.json:17]: Yes I think better to have it be P than P HH

[the remark this answers is not in the document]
Comment (joanise) [reply g2p/mappings/langs/generated/win-ipa_to_eng-ipa.json:17]: Glad to have your confirmation. I wasn't sure what made most sense in the languages.

**Review by joanise — APPROVED** [review]
Looks great!
